crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs

```diff
@@ -281,6 +306,100 @@ fn migrate_stores_to_v7(db: &IdbDatabase) -> Result<(), DomException> {
     db.delete_object_store(old_keys::INBOUND_GROUP_SESSIONS_V1)
 }
 
+async fn migrate_data_for_v8(name: &str, serializer: &IndexeddbSerializer) -> Result<()> {
+    // In migrate_data_for_v6, we incorrectly copied the keys in inbound_group_sessions verbatim into
+    // inbound_group_sessions2. What we should have done is re-encrypt them using the new table
+    // name, so we fix them up here.
+
+    info!("IndexeddbCryptoStore upgrade data -> v8 starting");
+
+    let db = IdbDatabase::open(name)?.await?;
+    let txn = db.transaction_on_one_with_mode(
+        keys::INBOUND_GROUP_SESSIONS_V2,
+        IdbTransactionMode::Readwrite,
+    )?;
+
+    let store = txn.object_store(keys::INBOUND_GROUP_SESSIONS_V2)?;
+
+    let row_count = store.count()?.await?;
+    info!(row_count, "Fixing inbound group session data keys");
+
+    // Iterate through all rows
+    if let Some(cursor) = store.open_cursor()?.await? {
+        let mut idx = 0;
+        let mut updated = 0;
+        let mut deleted = 0;
+        loop {
+            idx += 1;
+
+            // Get the old key and session
+
+            let old_key = cursor.key().ok_or(matrix_sdk_crypto::CryptoStoreError::Backend(
+                "inbound_group_sessions2 cursor has no key".into(),
+            ))?;
+
+            let idb_object: InboundGroupSessionIndexedDbObject =
+                serde_wasm_bindgen::from_value(cursor.value())?;
+            let pickled_session =
+                serializer.deserialize_value_from_bytes(&idb_object.pickled_session)?;
+            let session = InboundGroupSession::from_pickle(pickled_session)
+                .map_err(|e| IndexeddbCryptoStoreError::CryptoStoreError(e.into()))?;
+
+            if idx % 100 == 0 {
+                debug!("Migrating session {idx} of {row_count}");
+            }
+
+            // Work out what the key should be.
+            // (This is much the same as in `IndexeddbCryptoStore::get_inbound_group_session`)
+            let new_key = serializer.encode_key(
+                keys::INBOUND_GROUP_SESSIONS_V2,
+                (&session.room_id, session.session_id()),
+            );
+
+            if new_key != old_key {
+                // We have found an entry that is stored under the incorrect old key
+
+                // Delete the old entry under the wrong key
+                cursor.delete()?;
+
+                // Check for an existing entry with the new key
+                let new_value = store.get(&new_key)?.await?;
+
+                // If we found an existing entry, it is more up-to-date, so we don't need to do
+                // anything more.
+
+                // If we didn't find an existing entry, we must create one with the correct key
+                if new_value.is_none() {
+                    // This is much the same as `IndexeddbStore::serialize_inbound_group_session`.
+                    let new_data =
```

Comment:
can't we just use the `idb_object` we already have here?

Reply:
Great, thanks. Done in f81fb615e0